packages/firebase_auth/firebase_auth_platform_interface/lib/src/method_channel_firebase_auth.dart

```diff
@@ -0,0 +1,459 @@
+      <String, String>{'provider': provider, 'app': app},
+    );
+  }
+
+  @override
+  Future<void> updateEmail(String app, String email) {
+    return channel.invokeMethod<void>(
+      'updateEmail',
+      <String, String>{'email': email, 'app': app},
+    );
+  }
+
+  @override
+  Future<void> updatePhoneNumber(
+    String app,
+    PhoneAuthCredential phoneAuthCredential,
+  ) {
+    return channel.invokeMethod<void>(
+      'updatePhoneNumberCredential',
+      <String, dynamic>{
+        'app': app,
+        'provider': phoneAuthCredential.providerId,
+        'data': phoneAuthCredential.asMap(),
+      },
+    );
+  }
+
+  @override
+  Future<void> updatePassword(String app, String password) {
+    return channel.invokeMethod<void>(
+      'updatePassword',
+      <String, String>{'password': password, 'app': app},
+    );
+  }
+
+  @override
+  Future<void> updateProfile(
+    String app, {
+    String displayName,
+    String photoUrl,
+  }) {
+    final Map<String, String> arguments = <String, String>{'app': app};
+    if (displayName != null) {
+      arguments['displayName'] = displayName;
+    }
+    if (photoUrl != null) {
+      arguments['photoUrl'] = photoUrl;
+    }
+    return channel.invokeMethod<void>(
+      'updateProfile',
+      arguments,
+    );
+  }
+
+  @override
+  Future<void> setLanguageCode(String app, String language) {
+    return channel.invokeMethod<void>('setLanguageCode', <String, String>{
+      'language': language,
+      'app': app,
+    });
+  }
+
+  @override
+  Stream<PlatformUser> onAuthStateChanged(String app) {
+    Future<int> _handle;
+
+    StreamController<PlatformUser> controller;
+    controller = StreamController<PlatformUser>.broadcast(onListen: () {
+      _handle = channel.invokeMethod<int>('startListeningAuthState',
+          <String, String>{"app": app}).then<int>((dynamic v) => v);
+      _handle.then((int handle) {
+        _authStateChangedControllers[handle] = controller;
+      });
+    }, onCancel: () {
+      _handle.then((int handle) async {
+        await channel.invokeMethod<void>("stopListeningAuthState",
+            <String, dynamic>{"id": handle, "app": app});
+        _authStateChangedControllers.remove(handle);
+      });
+    });
+
+    return controller.stream;
+  }
+
+  @override
+  Future<void> verifyPhoneNumber(
+    String app, {
+    @required String phoneNumber,
+    @required Duration timeout,
+    int forceResendingToken,
+    @required PhoneVerificationCompleted verificationCompleted,
+    @required PhoneVerificationFailed verificationFailed,
+    @required PhoneCodeSent codeSent,
+    @required PhoneCodeAutoRetrievalTimeout codeAutoRetrievalTimeout,
+  }) {
+    final Map<String, dynamic> callbacks = <String, dynamic>{
+      'PhoneVerificationCompleted': verificationCompleted,
+      'PhoneVerificationFailed': verificationFailed,
+      'PhoneCodeSent': codeSent,
+      'PhoneCodeAuthRetrievalTimeout': codeAutoRetrievalTimeout,
```

Comment:
I think this 4 keys should be extracted to `const String` so they can be reused in the `_callHandler` method later.

Reply:
(Or maybe a small object that groups all these 4 handlers internally with some setters/getters so we don't show the Map at all, except when we need it here)

Reply:
Agreed, having a typed utility class is much better than an untyped `Map`